File: 07-Repeticion/03-Seguridad/04-Javascript/Seguridad.js
/*
   Crear un programa para contar cuántas letras mayúsculas,
   cuántas minúsculas y cuántos dígitos contiene una contraseña

   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Septiembre 2025
   Licencia: GNU GPL v3
*/

function main() {
	let contrasenia = ingresarTexto('contrasenia'); 

	let cantidadMayusculas = contarMayusculas(contrasenia);
	let cantidadMinusculas = contarMinusculas(contrasenia);
	let cantidadDigitos = contarDigitos(contrasenia);

	let reporteEstres = generarReporteContrasenia(contrasenia,
		cantidadMayusculas,cantidadMinusculas, cantidadDigitos);

	mostrarMensaje(reporteEstres);
}

function contarMayusculas (texto) {
	return contarCaracteres(texto, esMayuscula);
} 

function contarMinusculas  (texto) {
	return contarCaracteres(texto, esMinuscula);
}

function contarDigitos     (texto) {
	return contarCaracteres(texto, esDigito);
}

function esMayuscula (caracter) { return /\p{Lu}/u.test(caracter); }
function esMinuscula (caracter) { return /\p{Ll}/u.test(caracter); }
function esDigito    (caracter) { return /\p{Nd}/u.test(caracter); }

function contarCaracteres(texto, funcion) {
    let cantidadCaracteres = 0;

	for (let i = 0 ; i < texto.length ; i++) {
		if (funcion(texto.charAt(i))) {
			cantidadCaracteres += 1
		}
	}
	
	return cantidadCaracteres;
}

function generarReporteContrasenia(contrasenia,
	cantidadMayusculas, cantidadMinusculas, cantidadDigitos) {
	return `\nEn la constraseña \"${contrasenia}\"  hay:\n ` +
		   `${cantidadMayusculas} mayúscula(s), ` +
		   `${cantidadMinusculas} minúscula(s) y ` +
		   `${cantidadDigitos} dígito(s).\n`;
}

File: 07-Repeticion/03-Seguridad/03-Java/Seguridad.java
/*
   Crear un programa para contar cuántas letras mayúsculas,
   cuántas minúsculas y cuántos dígitos contiene una contraseña

   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Septiembre 2025
   Licencia: GNU GPL v3
*/

import modulo.Util;

void main() {
	var contrasenia = Util.ingresarTexto("Ingrese la contraseña analizar: ");

	var cantidadMayusculas = contarMayusculas(contrasenia);
	var cantidadMinusculas = contarMinusculas(contrasenia);
	var cantidadDigitos = contarDigitos(contrasenia);

	var reporteContrasenia = generarReporteContrasenia(contrasenia,
		cantidadMayusculas, cantidadMinusculas,	cantidadDigitos);

	Util.mostrarMensaje(reporteContrasenia);
}

int contarMayusculas(String texto)  { 
	return contarCaracteres(texto, esMayuscula); 
}

int contarMinusculas(String texto)  { 
	return contarCaracteres(texto, esMinuscula); 
}

int contarDigitos(String texto)     { 
	return contarCaracteres(texto, esDigito); 
}

IntPredicate esMayuscula = Character::isUpperCase;
IntPredicate esMinuscula = Character::isLowerCase;
IntPredicate esDigito    = Character::isDigit;

int contarCaracteres(String texto, IntPredicate funcion) {
	var cantidadCaracteres = 0;

	for (var caracter : texto.toCharArray())  {
		if (funcion.test(caracter)) {
			cantidadCaracteres++;
		}
	}

	return cantidadCaracteres;
}

String generarReporteContrasenia(String contrasenia, 
int cantidadMayusculas, int cantidadMinusculas, int cantidadDigitos) {
	return String.format("\nEn la constraseña \"%s\" hay:\n"+
		"%d mayúscula(s), "+
		"%d minúscula(s) y "+
		"%d dígito(s)\n",
		contrasenia, cantidadMayusculas, cantidadMinusculas, cantidadDigitos);
}
